File: Python/best_time_to_buy_and_sell_stock_ii.py
def maxProfit_greed(prices: list) -> int:
    length = len(prices)

    if length == 0:
        return 0
    
    profit = 0
    for i in range(1, length):
        profit = profit + max(prices[i] - prices[i - 1], 0)
    
    return profit

def maxProfit_dp(prices: list) -> int:
    length = len(prices)

    if length == 0:
        return 0
    
    profit1 = 0
    profit2 = -prices[0]
    for i in range(1, length):
        temp = profit1
        profit1 = max(profit1, profit2 + prices[i])
        profit2 = max(profit2, temp - prices[i])
    
    return profit1


if __name__ == "__main__":
    print(maxProfit_dp([7,1,5,3,6,4]))

File: Java/src/com/dbc/algorithm/Daily/BestTimeToBuyAndSellStockII.java
package com.dbc.algorithm.Daily;

public class BestTimeToBuyAndSellStockII {
    public static int maxProfitDp(int[] prices) {
        int length = prices.length;
        if (length == 0)
            return 0;

        int profit1 = 0;
        int profit2 = -prices[0];
        for (int i = 1; i < length; i++) {
            int temp = profit1;
            profit1 = Math.max(profit1, profit2 + prices[i]);
            profit2 = Math.max(profit2, profit1 - prices[i]);
        }

        return profit1;
    }

    public static int maxProfitGreed(int[] prices) {
        int length = prices.length;
        if (length == 0)
            return 0;

        int profit = 0;
        for (int i = 1; i < length; i++) {
            profit += Math.max(prices[i] - prices[i - 1], 0);
        }
        return profit;
    }
}
